File: abelkhan_type_csharp_component_msgpack/component/center/hubmanager.cs
/*
 * hubmanager
 * 2020/6/2
 * qianqians
 */

using System;
using System.Collections.Generic;

namespace abelkhan
{
    public class hubproxy
    {
        public string type;
        public string hub_type;
        public string name;
        public bool is_closed;
        private abelkhan.center_call_hub_caller _center_call_hub_caller;

        public hubproxy(abelkhan.Ichannel ch, abelkhan.modulemng modules, string _type, string _hub_type, string _name)
        {
            type = _type;
            hub_type = _hub_type;
            name = _name;
            _center_call_hub_caller = new abelkhan.center_call_hub_caller(ch, modules);
        }

        public void distribute_server_address(string type, string name, string ip, ushort port)
        {
            _center_call_hub_caller.distribute_server_address(type, name, ip, port);
        }

        public void reload()
        {
            _center_call_hub_caller.reload();
        }
    }

    public class hubmanager
    {
        private abelkhan.modulemng modules;
        private Dictionary<abelkhan.Ichannel, hubproxy> hubproxys;

        public hubmanager(abelkhan.modulemng _modules)
        {
            modules = _modules;
            hubproxys = new Dictionary<abelkhan.Ichannel, hubproxy>();
        }

        public hubproxy reg_hub(abelkhan.Ichannel ch, string _type, string _hub_type, string _name)
        {
            var _hubproxy = new hubproxy(ch, modules, _type, _hub_type, _name);
            hubproxys.Add(ch, _hubproxy);
            return _hubproxy;
        }

        public hubproxy get_hub(abelkhan.Ichannel ch)
        {
            if (hubproxys.TryGetValue(ch, out hubproxy _proxy))
            {
                return _proxy;
            }

            return null;
        }

        public hubproxy find_hub(string name)
        {
            foreach (hubproxy _proxy in hubproxys.Values)
            {
                if (_proxy.name.Equals(name)) {
                    return _proxy;
                }
            }

            return null;
        }

        public void for_each_hub(Action<hubproxy> fn)
        {
            foreach (var _proxy in hubproxys.Values)
            {
                fn(_proxy);
            }
        }

        public void hub_closed(abelkhan.Ichannel ch)
        {
            if (hubproxys.TryGetValue(ch, out hubproxy _proxy))
            {
                _proxy.is_closed = true;
            }
        }

        public bool check_all_hub_closed()
        {
            bool _all_closed = true;
            foreach (var _proxy in hubproxys.Values)
            {
                if (!_proxy.is_closed)
                {
                    _all_closed = false;
                }
            }
            return _all_closed;
        }
    }
}

File: abelkhan_type_component/component/center/hubmanager.ts
import abelkhan = require('../../abelkhan_type/ts/abelkhan');
import center_protcol = require('../../protcol/ts/center');

export class hubproxy{
    public is_closed : boolean = false;

    private center_call_hub_caller : center_protcol.center_call_hub_caller;
    constructor(ch:abelkhan.Ichannel, modules:abelkhan.modulemng){
        this.center_call_hub_caller = new center_protcol.center_call_hub_caller(ch, modules);
    }

    public distribute_server_address(type:string, name:string, ip:string, port:number){
        this.center_call_hub_caller.distribute_server_address(type, name, ip, port);
	}

    public reload(){
        this.center_call_hub_caller.reload();
    }
}

export class hubmanager{
    private modules : abelkhan.modulemng;
    private hubproxys : Map<abelkhan.Ichannel, hubproxy>;
    
    constructor(modules:abelkhan.modulemng){
        this.modules = modules;
        this.hubproxys = new Map<abelkhan.Ichannel, hubproxy>();
    }

    public reg_hub(ch:abelkhan.Ichannel){
		let _hubproxy = new hubproxy(ch, this.modules);
		this.hubproxys.set(ch, _hubproxy);
		return _hubproxy;
    }
    
    public get_hub(ch:abelkhan.Ichannel){
        return this.hubproxys.get(ch);
    }

    public for_each_hub(fn:(_proxy:hubproxy)=>void){
        this.hubproxys.forEach((_proxy:hubproxy)=>{
            fn(_proxy);
        });
    }

    public hub_closed(ch:abelkhan.Ichannel){
        let _proxy = this.hubproxys.get(ch);
        if (_proxy){
            _proxy.is_closed = true;
        }
    }

    public check_all_hub_closed(){
        let _all_closed = true;
        this.hubproxys.forEach((_proxy:hubproxy)=>{
            if (!_proxy.is_closed){
                _all_closed = false;
            }
        });
        return _all_closed;
    }

}
